Only lead climbs

pectum83 opened the pull request:
Evolution liée au ticket "croix en tete et/ou en moulinette https://github.com/oblyk/oblyk-app/issues/59" pour les deux points.
1 - Permet de n'afficher que les croix en tête avec un selecteur pour ceux qui veulent voir aussi les moulinettes.
2 - Ne compte pas les répétitions d'une même voie dans les croix

Corrige également : 
- les grandes voies sont comptées comme croix dans le niveau max parmis les sections réalisées dans l'ascent (auparavant elles comptaient dans le niveau min de toutes les sections de la crag_route)
- la hauteur des sections non faites d'une grande voie etaient comptées dans le total grimpé

A combiner avec la pull request sur oblyk-app

pectum83 committed:
current_users_controller ne retourne plus que les voies faites en tête

pectum83 committed:
les charts ne retournent plus que les voies faites en tête

pectum83 committed:
add a boolean to return all climbs or only lead climbs

pectum83 committed:
ajout grandes voies et type_chart

pectum83 committed:
compte les grandes voies dans le niveau de la section la plus dure realisée au lieu du niveau de la section la plus facile.

pectum83 committed:
Ce compte pas les répétitions d'une même voie dans les croix

pectum83 committed:
Ne compte pas les répétitions d'une même voie dans les croix

pectum83 committed:
Ne pas compter les sections non faites d'une GV dans le total de hauteur grimpée

pectum83 committed:
Ne pas compter les sections non faites d'une GV dans le total de hauteur grimpée

pectum83 committed:
Merge remote-tracking branch 'fork_api/onlyLeadClimbs' into onlyLeadClimbs

pectum83 committed:
correction gitignore

lucien-chastan commented on app/controllers/api/v1/log_books/outdoors_controller.rb:
```diff
@@ -8,27 +8,33 @@ class OutdoorsController < ApiController
         def figures
-          render json: LogBook::Outdoor::Figure.new(@user).figures, status: :ok
+          only_lead_climbs = params.fetch(:only_lead_climbs)
+          render json: LogBook::Outdoor::Figure.new(@user).figures(only_lead_climbs), status: :ok
```
on va plutôt faire un paramètre `filters` qu'on passera dans le l'initialiser de `LogBook::Outdoor::Figure.new`
Par exemple : 
```ruby
filters = {
  only_lead_climbs: params.fetch(:only_lead_climbs, 'false')
}
LogBook::Outdoor::Figure.new(@user, filters).figures
```
Ça permettra d'ajouter des filtres plus facilement par la suite.

en suite pour éviter de refaire le `filters` qui récupère le même paramètre dans chaque def, il faudrait un `before_action :set_filters` de la même manière que le `set_user`

lucien-chastan commented on app/models/ascent.rb:
```diff
@@ -18,6 +18,7 @@ class Ascent < ApplicationRecord
   validates :ascent_status, inclusion: { in: AscentStatus::LIST }
 
   scope :made, -> { where.not(ascent_status: :project) }
+  scope :lead, -> { where(roping_status: [:lead_climb, :multi_pitch_leader, :multi_pitch_alternate_lead]) }
```
histoire de compléter la méthode, on va aussi faire un scope `second` qui de la même manière récupérer les moulinettes et les seconds de cordé

pectum83 replied:
"Active Record already defines a method named `second`, which is used to fetch the second record from a query (similar to `first`)."  ==> je la nomme donc noLead

lucien-chastan commented on app/models/log_book/outdoor/chart.rb:
```diff
@@ -9,7 +9,15 @@ def initialize(user)
       end
 
-      def climb_type
+      def uniq_ascent_crag_routes(only_lead_climbs=false)
```
Du coup cette méthode ne va pas prendre de paramètre
Elle se basera sur le paramètre @filters passé à l'initialisation : )

et on peut déplacer cette méthode dans la section priver, elle n'a pas trop de raison d'être exposée

lucien-chastan commented on app/models/log_book/outdoor/figure.rb:
```diff
@@ -9,41 +9,58 @@ def initialize(user)
 
       def crags_count
-        @user.ascended_crags.distinct.count
+        @user.ascended_crags.distinct.count(:name)
```
Ha oui il y a peut-être un bug ici
Mais du coup on peut compter les (:id) parce qu'il y a des falaises qui on le même nom

lucien-chastan reviewed (commented):
Hello @pectum83 
Encore désolé pour le temps de relecture : /
Voici mes commentaires, n'hésite pas si tu as des questions

pectum83 committed:
Merge branch 'refs/heads/master' into onlyLeadClimbs

# Conflicts:
#	app/controllers/api/v1/current_users_controller.rb

pectum83 committed:
Refactor only_lead_climbs parameter handling

Updated the handling of the 'only_lead_climbs' parameter across various controllers to ensure consistent behavior by casting the parameter to boolean using ActiveModel. Adjusted schema character set from utf8mb4 to utf8mb3 for multiple tables, optimizing compatibility. Added debug logs and comments within methods related to climbing figures and charts to facilitate future debugging and enhancements.

pectum83 committed:
fix(ascent height = nil)

pectum83 committed:
Refactor crag route retrieval in chart and figure models

pectum83 committed:
Add 'second' scope to ascent model

Introduce 'second' scope to filter ascents based on top rope and multi-pitch second roping statuses. This enhancement improves query flexibility for retrieving specific ascent types.

pectum83 committed:
Refactor ascent filtering and chart generation

Introduce a unified filter system using `CragAscentFilter` for ascent data, replacing individual filtering logic in controllers and models. This simplifies the codebase, enhances reusability, and improves maintainability of climbing data filters and analytics.

pectum83 committed:
Refactor filtering logic and clean up ascent-related APIs

Streamline and centralize filtering logic through the new `CragAscentFilters` class, replacing scattered filter handling. Simplify API endpoints, models, and services by delegating responsibilities to the updated filtering mechanism, improving maintainability and performance.

pectum83 committed:
Refine comments

pectum83 committed:
use filter all figures (country, region, crags)

pectum83 committed:
Add filter for user_controller (other user than connected one)

pectum83 commented on db/schema.rb:
je ne comprend pas quand ce fichier a changé mais je n'arrive pas a le retirer de la PR. Dis moi si ca ne fusionne pas et je réessaierai.

pectum83 committed:
clean unmodified files

pectum83 committed:
Revert "correction gitignore"

This reverts commit 10b0e8e203da1b5941b21dc4e582a9fc5e3a231a.

pectum83 committed:
supress gitignore

pectum83 commented on .gitignore:
je n'arrive pas a supprimer le .gitignore. Dis moi si ca pose problème et si tu sais comment faire je réessaierai.

pectum83 commented:
> Hello @pectum83 Encore désolé pour le temps de relecture : / Voici mes commentaires, n'hésite pas si tu as des questions

pectum83 commented:
Salut @lucien-chastan,

Désolé pour les fausses manip sur la PR mais je n'ai pas l'habitude de gilthub. J'espère que ça ne t'as pas trop spammé.

J'ai intégré tes remarques précédentes. L'intégration des climbing_type en filters a finalement conduit a pas mal de modif. Au passage comme je devais refactoré pas mal de choses j'en ai profité pour homogénéiser certaines parties qui faisaient la même chose de deux manières différentes. J'espère que je n'ai rien cassé. J'ai fait pas mal de tests mais ma base de donnée est toute petite et je ne vois pas les éventuels problèmes de latence.

Les points qui restent en attente (et que je ferai avant merge selon ce que tu décides):
- je n'ai pas mis le button toggle de filtre dans la vue des "autres" users. Mais je peux si tu préfères
- j'ai mis le résultat des requetes de filtre dans un cache court de 30 sec (le temps de charger les différents widgets de la page) mais je ne sais pas si ca sera compatible de l'utilisation massive et de ton serveur. Il y a un bool en dur (@use_cache = true) dans crag_ascent_filters pour ne pas faire de cache. Mais si tu préfère je retire le cache.

Vue l'ampleur des changements je pense qu'il faudra une autre passe de commentaires et de reprise de code pour converger vers quelque chose qui te corresponde. J'ai essayé d'être homogène avec le reste du code mais c'est pas évident de savoir comment tu aurais fait les choses.

N'hésites pas à faire des remarques et je modifierai.

Christophe.

pectum83 commented:
quelques screenshots
![Screenshot from 2024-12-12 14-26-17](https://github.com/user-attachments/assets/77838b59-2c1b-479d-a419-a04eb111cfab)
![Screenshot from 2024-12-12 14-26-00](https://github.com/user-attachments/assets/485e86d0-d974-42f4-9a66-029c2028949a)
![Screenshot from 2024-12-12 14-25-36](https://github.com/user-attachments/assets/5b6e250c-4610-4a19-990e-141054ff0c26)

pectum83 committed:
Refactor ascent filtering to prioritize best ascents.

Updated the filtering logic to include sorting by roping_status and ascent_status before removing duplicates. This ensures the "best" ascent for each crag route is retained when applying the no_double filter.

pectum83 commented on app/models/ascent.rb:
```diff
@@ -18,7 +18,31 @@ class Ascent < ApplicationRecord
+  scope :by_climbing_type, ->(type) { joins(:crag_route).where(crag_routes: { climbing_type: type }) }
+  # scope for uniqueness: keep only the last ascent for each crag_route_id ; base_query arg is for executing this one after the others on resulting ascents
+  # requete SQL probablement trop lente -> on utilise donc un Rails uniq dans le modele appelant
+  # TODO a supprimer apres confirmation avec Lucien
```
Je pense qu'il vaut mieux faire le 'no-double' filter en ruby (uniq) comme fait dans crag_ascent_filter.rb. Si tu confirmes je supprime ce filtre des scopes.

pectum83 commented on app/services/crag_ascent_filters.rb:
```diff
@@ -0,0 +1,94 @@
+  }.freeze
+
+  def initialize(user, params)
+    @use_cache = true # TODO decide here if use cache or not
```
Selon ton avis ou les essais sur la bdd complète, changer ici l'utilisation ou pas du cache (si on ne l'utilise pas je peux même supprimer tout le code de cache pour améliorer la lisibilité.

lucien-chastan replied:
Oui je vois le problème que tu veux patcher avec le cache, vu que chaque graph / tableau / chiffre est appelé par une requêtes différentes, la liste des ascensions filtrées est rejouées à chaque fois ...
Mais le cache c'est souvent du cache misère _(et il y en a pas mal sur oblyk 😅 notamment pour la génération des variantes d'images, que je suis en train de tenter de [patcher](https://github.com/oblyk/oblyk-api/pull/18))_ c'est a utiliser en dernier recours !

30 secondes c'est suffisamment long pour que je regarde mon carnet de croix, que j'ajoute une croix, que je revienne et que je ne la voie pas.

et ça peux être perturbant à l'usage, je navigue c'est rapide, puis c'est lent, puis de-nouveau rapide pendant 30 seconds, etc.

Autre argument, en prod Oblyk a 3 forkes, tes requêtes sont appelées simultanément, tu n'as pas la garantie que la requêtes 2 profite du cache de la requête 1, il y a des chances que ton filtre soit calculé X fois et que chacun écrase le cache généré précédemment.

Le mieux serait de faire une nouvelle route dans notre controller qui prend comme paramètre une liste de 'module' à récupérer, avec les filtres.

Par exemple on pourrait avoir une route dans `controlles/api/v1/log_books/outdoors_controller.rb` qui pourrait ressembler à ça :

```ruby
def stats
  # set filters, etc.
  stats = {}
  stats[:figures] = LogBook::Outdoor::Figure.new(@filters).figures if params[:stats][:figures]
  stats[:climb_types_chart] = LogBook::Outdoor::Chart.new(@filters).climb_type if params[:stats][:climb_types_chart]
  stats[:ascended_crag_routes] = LogBook::Outdoor::List.new(@filters).ascents if params[:stats][:ascended_crag_routes]
  # etc.
  render json: stats, status: :ok
end
```
_(il faudrait mutualisé les new(), pour ne pas les instancier plusieurs fois)_

je me tâte sur le nom 'stats' de la route, peut-être que 'widgets' serait plus pertinent pour comprendre qu'on récupérer différent morceau du carnet de croix.

et `LogBook::Outdoor::List.new(@filters).ascents` n'hésite pas 🤣 mais en vrai ça serait bien que le code de ascended_crag_routes ne soit plus dans le controller de current_users_contoller.rb

une route telle que celle-ci permettrait de récupérer "à la carte" les différents modules du carnet de croix dont on a besoin en une seul requête et éviterait de passer par un cache du filtre

Et sur le front ça ne change pas grand chose, dans `pages/home/ascents/outdoor/index.vue` par exemple, on remplace les 3 appels à API par un qui récupère simultanément les data de `figures`, `climbingTypeData` et `gradeData`

Mais pour répondre à ta question, je préfère qu'il n'y ai pas de cache dans ce module 😅

pectum83 replied:
J'ai tout factoriser concernant le logBook en intégrant current_user et user controllers dans outdoor controller. Je pense que pour plus de cohérence il faudrait du coup déplacer les routes logbooks/outdoors a la racine, en dehors de currentuser. Qu'est ce que tu en penses ?

lucien-chastan commented:
Hello @pectum83 !

Quand je vois ce développement je me dis qu'il y a quelque chose à faire !

Mais qui va demander un peu plus de travail 😬

En fait avoir un bouton 'uniquement à vue' je me dis : "et si je veux voir uniquement mes flashs ?"
Idem pour "sans moulinette", ça me kick toujours de sélectionner une négation.

Un autre sujet, et c'est là où ça va plus loin, depuis un moment je me dis que c'est dommage de pouvoir choisir le type d'escalade (bloc, voie, grand voie, etc.) uniquement sur le liste, et par sur les chiffres ni sur Analyticks.

Donc je te propose d'aller un peu plus loin 😅

Voici ce que j'ai en tête :

![image](https://github.com/user-attachments/assets/14f434de-2d31-487e-9cac-5364020dd066)

Avant les onglets, on ajoute une boîte que l'on peu déplier avec le bouton (FILTRER MON CARNET)
ça déploie la boîte et laisse choisir parmi "le type de grimpe / le status / l'encordement" de choisir un ou plusieurs items
et on soumet le filtre avec le bouton 'Filtrer'

Ce qui serait bien c'est que les filtres soit passé en paramètres des liens des onglets en dessous, pour garder les filtrer entre les différentes page.

Tu en pense quoi ?

C'est fait vite fait, il y a des choses à améliorer en terme d'affichage, c'est dans l'esprit

_(ça risque de poser des questions sur la page 'Tick list', 'Projet' et "ma carte" qui sont un peu différentes ...)_

pectum83 commented:
Ca me plait, je regarde ça. 
Pour l'instant les filtres sont déjà échangés avec l'onglet Analytik au travers du LocalStorage. Ca permet aussi de conserver le filtre préféré entre chaque session.
J'ajouterai bien aussi un bouton reset pour remettre tous les filtres off d'un clic.

lucien-chastan commented:
Cool !

Le LocalStorage ça marche aussi !

Le bouton reste bonne idée 👍

Désolé de rajouter des choses à faire sur cette PR que j'ai mis temps de temps à relire !

pectum83 committed:
Refactor ascent filtering

pectum83 commented:
Voila une nouvelle version generalisée des filtres. N'hesites pas à me dire si tu veux des amélioration d'aspect ou de confirmité de code.
![Screenshot from 2024-12-27 09-21-20](https://github.com/user-attachments/assets/0a4dbd94-fdea-4175-bfc2-4cbea298d64a)
![Screenshot from 2024-12-27 09-23-42](https://github.com/user-attachments/assets/f3e578e6-69e6-40b5-9cdb-d256e482d8c3)
![Screenshot from 2024-12-27 09-23-51](https://github.com/user-attachments/assets/cba85df9-4be3-4270-bfa5-ed0b83514d03)
![Screenshot from 2024-12-27 09-24-15](https://github.com/user-attachments/assets/42a08e07-37ef-4556-a280-010e1500490d)
![Screenshot from 2024-12-27 09-24-23](https://github.com/user-attachments/assets/64b05dc0-8439-4a74-8a2a-a7dc8d6b3372)

lucien-chastan commented on app/controllers/api/v1/current_users_controller.rb:
```diff
@@ -477,6 +468,10 @@ def set_user
       end
 
+      def set_filters
+        @filters = CragAscentFilters.new(@user, params)
```
c'est bien de filtrer les params avant de les envoyer dans un module

il faudrait ajouter quelque choses comme : 

```ruby

private

def filter_params
  # filtre des params
end
```
ça permet aussi d'expliciter ce qu'on attends comme paramètre : )

pectum83 replied:
voir mon comment sur le passage en POST au lieu de GET pour gérer  les param nested.

lucien-chastan replied:
Tu ne peux pas passer de data à une requête GET, mais tu peux passer des params

axios va transformer tes nested attributes, exemple
```js
params: {
  roping_status: ['lead', 'second']
} 
```
ça va être transformé en ma-route.json?roping_status[]=lead&roping_status[]=second

et ça params de ruby le comprend bien

par contre axios ne parse pas correctement les objects, est c'est bien dommage, parce que params de rails peut le comprendre

Est-ce qu'on a besoin de paramètre imbriqués ? autre que des tableaux ?

pectum83 replied:
si on met les filtres à la racine de params effectivement on peut garder un GET. Mais le problème c'est pour rassembler tous les filtres dans un param filters comme tu le souhaites.
```
params: {
  filters: {
    roping_status: ['lead', 'second']
    ascent_status: ...
  }
}
``` 
Et ça je n'ai pas réussi à la faire parser automatiquement par rails.
Mais si tu préfaire un GET avec les filtres à la racine, ça j'ai réussi. 
Dis moi ce que tu préfères ?

lucien-chastan replied:
Ha oui désolé, je trouvais ça plus propres de mettre dans une clé filters, mais je préfères que ça reste un GET : )
Donc on va oublier la clé filter 😬 ... et mettre à la racine
Désolé pour les aller retour ...

pectum83 replied:
C'était un peu lourd et verbeux de tout remettre a la racine (pour les filters mais aussi pour les stats_list). J'ai donc laissé les params nested et je les reparse à la main dans le back. Finalement ce n'est pas si verbeux que ça dans le back (Tu me diras ce que tu en  penses).
```
 filters_params = JSON.parse(params[:filters]).transform_keys(&:to_sym)
 filters = ActionController::Parameters.new(filters_params).permit(
          ascent_status_list: [],
          roping_status_list: [],
          climbing_types_list: []
        )
```

lucien-chastan commented on app/services/crag_ascent_filters.rb:
```diff
@@ -0,0 +1,85 @@
+# app/services/filter.rb
```
ligne à supprimer

par contre tu peux ajouter `# frozen_string_literal: true` et un retour à la ligne !

lucien-chastan commented on app/services/crag_ascent_filters.rb:
```diff
@@ -0,0 +1,85 @@
+  # but in Ruby (eg no_double filter with a uniq method)
+  # for now, only "no_double" filter is in this list
+  DEFAULTS_FILTERS = {
+    "ascentStatusList" => AscentStatus::LIST, # ['sent', 'red_point', 'flash', 'repetition']
```
convention, utiliser la notation my_key: my_value en snake case

donc :

```ruby
DEFAULTS_FILTERS = {
  ascent_status_list: AscentStatus::LIST,
  roping_status_list: RopingStatus::LIST,
  # etc.
}.freeze
```

lucien-chastan commented on app/services/crag_ascent_filters.rb:
```diff
@@ -0,0 +1,85 @@
+    @use_cache = true # TODO decide here if use cache or not
+    @user = user
+    @filters = filters_from_params(params)
+    Rails.logger.info("filters: #{@filters}") # TODO remove this line
```
so remove this line 🤓

Si tu veux inspecter ce qu'il se passe dans ton code, n'hésite surtout pas à abuser du `binding.pry` si tu ne connais pas déjà

lucien-chastan commented on app/services/crag_ascent_filters.rb:
```diff
@@ -0,0 +1,85 @@
+      @user.ascent_crag_routes.made.filtered(@filters).to_a
+    end
+  end
+
```
ligne à supprimer

lucien-chastan commented on app/services/crag_ascent_filters.rb:
```diff
@@ -0,0 +1,85 @@
+  def filters_from_params(params)
+    # Parse `filters` if it’s a string
+    filters = params[:filters]
+    filters = JSON.parse(filters) if filters.is_a?(String)
```
Il est préférable que les modules soit le plus stricte possible sur ses paramètres d'entrés.
Pour moi le module il attend un hash de paramètre, et ce n'est pas à lui de se tordre pour s'exécuter peut importe ce que tu lui passe

pectum83 replied:
Après avoir un peu tout essayé j'ai fini par comprendre que un GET avec des pramètres Nested n'est pas parser au rang deux par Ruby. Donc soit on parse a la main comme j'avais fait mais c'est un peu lourd et ca ne permet plsu de faire le require et permit sur les params. Soit on fait un POST (qui admet des data nested compatibles avec le parseur Ruby) a la place du GET. C'est ce que j'ai fait pour l'instnat dans la version que je vais commiter

lucien-chastan commented on app/services/crag_ascent_filters.rb:
Le dossier services et plus pour l'appel à des services extérieurs, telle que Google Elevation API, OpenStreetMapAPI, etc.

Je pense que cette class trouverait plus ça place dans app/models/log_book/outdoor

Pour le nom de la class il pourrait être plus parlant, en l'état on dirai que ça va juste nous renvoyer un hash bien formaté des filtres applicables sur la liste des ascents, et nom pas une liste d'ascension filtrées

Je pense que CragAscent est suffisant, et puisque qu'on lui passe un `@user`, et des `@filters`, on comprends que ça sera la liste des acensions filtré d'un utilisateur donnée

lucien-chastan commented on app/models/ascent.rb:
```diff
@@ -18,7 +18,22 @@ class Ascent < ApplicationRecord
   validates :ascent_status, inclusion: { in: AscentStatus::LIST }
 
   scope :made, -> { where.not(ascent_status: :project) }
+  scope :no_repetition, -> { where.not(ascent_status: :repetition) }
```
:no_repetition, :lead, et :on_sight ne sont plus utilisé ? On peut les supprimer ?

lucien-chastan commented on app/models/ascent.rb:
```diff
@@ -18,7 +18,22 @@ class Ascent < ApplicationRecord
+  # Combine all filters in the `filtered` scope
+  scope :filtered, ->(filters) {
+    scoped_results = self
+    scoped_results = scoped_results.by_ascent_statuses(filters["ascentStatusList"])
```
ça sera à changer en `filters[:ascent_status_list]`

lucien-chastan commented on app/models/log_book/outdoor/chart.rb:
```diff
@@ -3,71 +3,38 @@
-      def initialize(user)
-        @user = user
+      # filters parameter is an instance of CragAscentFilters
+      def initialize(filters)
```
je préfère que cette variable s'appelle 'ascents' plutôt que 'filters' (même raison que mon commentaire sur le nom du fichier crag_ascent_filters.rb)

idem pour les autres endroit : )

lucien-chastan commented on app/models/log_book/outdoor/chart.rb:
```diff
@@ -86,176 +53,87 @@ def grade
+            label: climbing_type
           }
-        end
+        end.compact
```
ça c'est vraiment très personnel, mais je trouve que les méthodes après les `end` c'est compliqué à lire
le `end` ça clôt les `class / module / def / if` dans tous les autres cas, donc on ne s'attend pas à voir une méthode après un `end`

je préfère voir un `datasets = datasets.compact` sur une nouvelle ligne

lucien-chastan commented on app/services/crag_ascent_filters.rb:
```diff
@@ -0,0 +1,85 @@
+  # others are applied in Ruby (eg no_double filter with a uniq method).
+  # filtered_ascents_array is an array of AscentCragRoute objects after applying all filters
+  # filtered_ascents_active_record is an ActiveRecord::Relation after applying only the filters that can be applied in the database query
+  attr_reader :filtered_ascents_array, :filtered_ascents_active_record
```
Ça m'embête un peu cette double manière de récupérer la liste des ascensions ...

C'est dans le cas ou l'on a plusieurs fois la même voie dans son carnet ?
Genre 'après travail' puis 'répétition', 'répétition', 'répétition' ?

pectum83 replied:
Ca m'embête aussi.
Le besoin de départ c'est de supprimer les doubles dans les stats (pas grave dans les listes par date ou crag mais dommage de compter la meme voie plusieurs fois dans les stats). Pour les répétitions ce n'est pas un problème, on peut les filtrer dans le filtre ascent_status. Le problème c'est plutot pour une voie faite par ex : a vue en moul, puis apres travail en tete, ou enchainée (statut qui n'existe plus mais que j'ai perso beaucoup utilisé avant de voir repetition). Autre exemple : si je veux toutes mes voies en tête, il faut remettre repetition (sinon je n'ai pas celles que j'ai faites a vue en moul et en tete en repetition). Mais si j'ai fait plusieurs repetitions en tete, je la compte plusieurs fois.

Du coup je propose : 
- de supprimer le filtre no_double
- de ne laisser que le active record dans CragAscentFilter et de filtrer les doubles 
- de filtrer les doubles systématiquement directement dans chart.rb et figure.rb

Est ce que ça te va ?

lucien-chastan replied:
Yes ça me va,
On fera le filtre des doublons dans les charts et figures !

lucien-chastan commented on app/services/crag_ascent_filters.rb:
```diff
@@ -0,0 +1,85 @@
+    "ascentStatusList" => AscentStatus::LIST, # ['sent', 'red_point', 'flash', 'repetition']
+    "ropingStatusList" => RopingStatus::LIST, # ['lead_climb', 'top_rope', ...]
+    "climbingTypesList" => Climb::CRAG_LIST,  # ['sport_climbing', 'bouldering', ...]
+    "otherFilters" => ["no_double"]
```
pourquoi pas `no_double: false` directement à la racine de DEFAULT FILTERS ?
Surtout si on part sur un array, si j'ai un autre 'otherFitlter' qui aurait besoin d'une valeur, on va être bloqué

lucien-chastan commented on app/controllers/api/v1/log_books/outdoors_controller.rb:
```diff
@@ -6,37 +6,37 @@ module LogBooks
       class OutdoorsController < ApiController
         before_action :protected_by_session
         before_action :set_user
+        before_action :set_filters, except: [:daily_ascents, :ascents_of_crag]
```
plus parlant en set_ascents 
et du coup `@filters` serait plutôt `@ascents`
(voir commentaire sur le nom de carg_ascent_filters)

lucien-chastan reviewed (commented):
Merci pour le clean que tu as fait dans certaines parties du code où il y avait beaucoup de répétions !

pectum83 committed:
suppress otherFilters. Replaced by boolean no_double

pectum83 committed:
lists in snakecase

pectum83 committed:
refactor @filters to @ascents

pectum83 committed:
factorize stats api

pectum83 committed:
POST pour les nested params et factorisation user et currentuser api avec outdoor_controller api

pectum83 committed:
call stats with GET and parse nested params  in api

pectum83 committed:
suppress log console

pectum83 committed:
axios send correctly stringify nested parameters

pectum83 committed:
changement camelcase et param a la racine

lucien-chastan commented on app/models/ascent.rb:
```diff
@@ -19,15 +19,15 @@ class Ascent < ApplicationRecord
-  scope :by_ascent_statuses, ->(ascentStatusList) { where(ascent_status: ascentStatusList) }
-  scope :by_roping_statuses, ->(ropingStatusList) { where(roping_status: ropingStatusList) }
-  scope :by_climbing_types, ->(climbingTypesList) { joins(:crag_route).where(crag_routes: { climbing_type: climbingTypesList }) }
+  scope :by_ascent_statuses, ->(ascentFilter) { where(ascent_status: ascentFilter) }
```
du coup snake_case ici 😅

pectum83 replied:
Ah oui. C'est fait.

pectum83 committed:
another snake case

pectum83 commented on config/routes.rb:
```diff
@@ -127,12 +122,8 @@
           namespace :log_books do
             resources :outdoors, only: %i[] do
```
Dans la mesure on on a factorisé le log_books pour le current user et les users en ajoutant un parametre user_id, il serait plus propre de déplacer le namespace log_books ailleurs que dans current_user. Qu'en penses tu ?

lucien-chastan replied:
Je ferais ça dans un second temps, s'il y a besoin, c'est plus simple et plus sécurisé que tous ce qui touche à son compte passe par current_users

Par contre il va falloir la route
users/:user_uuid/log_books/outoors/stats sinon la vision d'un carnet de croix d'un autre grimpeur ne fonctionnera plus, ou alors il faut laisse les routes dans users/ si on ne fait pas la refacto

pectum83 replied:
En fait j’ai déjà factorisé le back en ajoutant le test du user dans outdoorcontroller (ligne 84). Du coup le front des users appelle outdoorcontroller avec le user en param. C’est pour ça que je proposais de sortir la route outdoor du current user.

lucien-chastan replied:
ha merde désolé je n'avais pas vu que tu avais refactorisé pour pouvoir appeler le carnet de croix d'un autre user depuis la route /current_users/log_books/outdoors/stats.json

Ça ne me semble pas être une bonne chose, ça brise la logique que current_user c'est forcément le user connecté, et users c'est les autres

Et maintenant que j'y pense ça serai étrange de pouvoir filtrer le carnet de croix d'une autre personnes, comme on ne peux pas voir toutes les infos des autres (comme la carte d'escalade, les tick list, l'analytics, etc.)

je préfère qu'il y ai une route dans users qui initialise le module de stats avec des filtres par défait qu'un user ne peux pas changer

En fait j'ai l'impression que la partie carnet de croix d'un autre user pourrais ne pas changer ?
Peut-être juste passer des filtres par défaut dans l'initialisation sans lancer la possibilité de les changer

pectum83 commented:
ok je vais supprimer les filtres sur les users.
Pour la logique current_user / user, ce que je proposais c'était que la logique reste dans le front mais que les deux pages appellent la meme api (LogBookOutdoorApi) et du coup de sortir cette api de current user (avec un param user). Ca m'a permis de tout factoriser et de supprimer tous les doublons qui avaient des petites différences tout en faisant la même chose. En fait c'est ma "maniaquerie" perso de ne jamais avoir plus de 10 lignes de code dupliqué (je l'ai meme dans mon linter et husky m'interdit de commiter ;) ). Ceci étant dit, ok pour suivre tes souhaits d'architecture. Je dupliquerai l'api stats dans users_controller avec les filtres en dur pour ne rien filtrer (ou tout simplement pas de filtre et direct un .made). ET du coup pas besoin de factorisation dans le front.

pectum83 committed:
separation et duplication entre user et current user

pectum83 commented:
C'est fait. J'ai juste laissé le filtre "climbing type" dans les autres users pour rester identique a ce qu'il y a vait avant. Pour les autres users je n'ai pas fait de local storage. Ca reset le filtre climbing types a tous les types à chaque refresh.
Et tout est séparé entre les api current_user et users.

lucien-chastan commented on app/controllers/api/v1/log_books/outdoors_controller.rb:
```diff
@@ -85,7 +81,16 @@ def ascents_of_crag
 
         def set_user
-          @user = @current_user
+            @user = @current_user
```
indentation

lucien-chastan commented on app/models/log_book/outdoor/crag_filtered_ascents.rb:
```diff
@@ -0,0 +1,35 @@
+        @ascents = @user.ascent_crag_routes.made.filtered(@filters)
+      end
+
+      private
```
retour à la ligne après private

lucien-chastan commented on app/models/log_book/outdoor/list.rb:
```diff
@@ -0,0 +1,46 @@
+
+      def ascended_crag_routes(page, order)
+        ascents = @ascents.joins(crag_route: :crag)
+                                  .includes(
```
.includes devrait être au même niveau d'indentation que .joins

lucien-chastan commented on app/models/log_book/outdoor/list.rb:
```diff
@@ -0,0 +1,46 @@
+                                      crag: { photo: { picture_attachment: :blob } },
+                                      photo: { picture_attachment: :blob }
+                                    },
+                                    )
```
devrait avoir une indentation de moins

lucien-chastan reviewed (commented):
idem que pour l'autre PR, 2 / 3 retour sur de la typo

mais on va être bon !

pectum83 committed:
minor changes

pectum83 commented:
Normalement ca devrait être tout bon api et app

pectum83 committed:
suppress uniq

pectum83 commented:
Dernière retouche : finalement j'ai supprimé les uniq(crag_id) dans les charts et les figures. Avec les filtres complets il suffit de ne pas selectionner "répetition" et "enchainé"

lucien-chastan commented:
Parfait !
Je pense merge ça aujourd'hui ou demain : )
